TESTS/mbed_drivers/unbuffered_Serial/gt_unbuffered_serial_test_env.cpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2019, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if DEVICE_SERIAL
+
+#include <string>
+
+#include "drivers/UnbufferedSerial.h"
+
+
+/**
+ * Macros for setting console flow control.
+ */
+#define CONSOLE_FLOWCONTROL_RTS                                               1
+#define CONSOLE_FLOWCONTROL_CTS                                               2
+#define CONSOLE_FLOWCONTROL_RTSCTS                                            3
+#define mbed_console_concat_(x) CONSOLE_FLOWCONTROL_##x
+#define mbed_console_concat(x) mbed_console_concat_(x)
+#define CONSOLE_FLOWCONTROL mbed_console_concat(MBED_CONF_TARGET_CONSOLE_UART_FLOW_CONTROL)
+
+
+/**
+ *  \brief Current token of key-value protocol's tokenizer
+ */
+static int cur_token = 0;
+
+/**
+ *  \enum Token enumeration for key-value protocol tokenizer
+ *
+ *        This enum is used by key-value protocol tokenizer
+ *        to detect parts of protocol in stream.
+ *
+ *        tok_eof       ::= EOF (end of file)
+ *        tok_open      ::= "{{"
+ *        tok_close     ::= "}}"
+ *        tok_semicolon ::= ";"
+ *        tok_string    ::= [a-zA-Z0-9_-!@#$%^&*()]+    // See isstring() function
+ *
+ */
+enum Token {
+    tok_eof = -1,
+    tok_open = -2,
+    tok_close = -3,
+    tok_semicolon = -4,
+    tok_string = -5
+};
+
+
+class GreenteaUnbufferedSerial : public mbed::UnbufferedSerial {
+public:
+    GreenteaUnbufferedSerial();
+};
+
+GreenteaUnbufferedSerial::GreenteaUnbufferedSerial() :
+    mbed::UnbufferedSerial(USBTX, USBRX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE)
+{
+#if   CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTS
+    set_flow_control(SerialBase::RTS, STDIO_UART_RTS, NC);
+#elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_CTS
+    set_flow_control(SerialBase::CTS, NC, STDIO_UART_CTS);
+#elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTSCTS
+    set_flow_control(SerialBase::RTSCTS, STDIO_UART_RTS, STDIO_UART_CTS);
+#endif
+}
+
+GreenteaUnbufferedSerial greentea_unbuffered_serial;
+
+
+static int greentea_read() {
+    int c = 0;
+    greentea_unbuffered_serial.read(&c, 1);
```

Comment:
You are currently limited with reading one byte at a time because the unbuffered serial class is constrained by parse_kv. You don't really need to parse key-value pair to test unbuffered serial because you can fix how much data you want to read/write and what data you want to read/write.

So you can get rid of thsi function and all the functions required to parse kv. Instead just call `test_unbuffered_serial.read(buffer, number_of_bytes_to_read)`

Reply:
Look at the implemention of `UnbufferedSerial::read` you cannot really do that. `UnbufferedSerial::read` only reads one char at a time as implementing read on a TTY `getc` you have to read exactly 1 character.

Reply:
I had not reviewed UnbufferedSerial yet. 
We cannot provide a read function that only reads one character at a time. The implemenattion must be changed.

Reply:
If `UnbufferedSerial` needs to work like `DirectSerial`, read [this](https://arm-isg.slack.com/archives/DD1UH3QEA/p1574328548000200).

Reply:
OK. UnbufferedSerial will return one character on a read.
So that means that your test needs to implement a read function around `test_unbuffered_serial.read` so it can read as many chars as we need to. 
I would write a read wrapper function in your main file to do that.

Reply:
Please review [this](https://github.com/hugueskamba/mbed-os/compare/310a7cd73ede0bcfff6c99b4b3ae1c891754b989..dd9ce703013101cd3ddc946a7e0f393357532b49).